### PR TITLE
Minimal integration with the Skia text shaper module

### PR DESCRIPTION
**Pull request by jason-simmons**
This converts the libtxt Paragraph and ParagraphBuilder classes into
interfaces with Minikin and SkShaper/SkParagraph based implementations.

Use the --enable-skshaper GN flag to select the Skia shaper implementation
at build time.

### CODE REVIEW

File: lib/ui/text/paragraph_builder.cc [hunk 1]
```diff
@@ -281,9 +281,15 @@ ParagraphBuilder::ParagraphBuilder(
 
   FontCollection& font_collection =
       UIDartState::Current()->window()->client()->GetFontCollection();
-  m_paragraphBuilder = std::make_unique<txt::ParagraphBuilder>(
+
+#if FLUTTER_ENABLE_SKSHAPER
```
Comment (gaaclarke) [inline lib/ui/text/paragraph_builder.cc]: I'd recommend reducing the amount of divergent code.
```
#if FLUTTER_ENABLE_SKSHAPER
#define FLUTTER_PARAGRAPH_BUILDER txt::ParagraphBuilder::CreateSkiaBuilder
#else
#define FLUTTER_PARAGRAPH_BUILDER txt::ParagraphBuilder::CreateTxtBuilder
#endif
m_paragraphBuilder = FLUTTER_PARAGRAPH_BUILDER(style, font_collection.GetFontCollection());
```

File: third_party/txt/src/skia/paragraph_builder_skia.cc [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+namespace skt = skia::textlayout;
+
+namespace txt {
+
+namespace {
+
+SkFontStyle MakeSkFontStyle(txt::FontWeight font_weight,
+                            txt::FontStyle font_style) {
+  return SkFontStyle(static_cast<SkFontStyle::Weight>(
+                         static_cast<int>(font_weight) * 100 + 100),
```
Comment (gaaclarke) [inline third_party/txt/src/skia/paragraph_builder_skia.cc]: I'd pull out the magic numbers into named variables to better document them.

File: third_party/txt/src/skia/paragraph_builder_skia.cc [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  skt::TextStyle text_style;
+
+  text_style.setFontStyle(MakeSkFontStyle(txt.font_weight, txt.font_style));
+  text_style.setFontSize(SkDoubleToScalar(txt.font_size));
+  text_style.setHeight(SkDoubleToScalar(txt.height));
+  text_style.setFontFamilies({SkString(txt.font_family.c_str())});
+  text_style.setLocale(SkString(txt.locale.c_str()));
+  skia.setTextStyle(text_style);
+
+  // TODO: strut
```
Comment (gaaclarke) [inline third_party/txt/src/skia/paragraph_builder_skia.cc]: Shouldn't the TODO comment be a full sentence?

File: third_party/txt/src/skia/paragraph_builder_skia.cc [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+  // General paragraph properties.
+  skia.setTextAlign(static_cast<skt::TextAlign>(txt.text_align));
+  skia.setTextDirection(static_cast<skt::TextDirection>(txt.text_direction));
+  skia.setMaxLines(txt.max_lines);
+  skia.setEllipsis(txt.ellipsis);
+
+  // TODO: break strategy?...
+
+  // This is something flutter expect us to do
```
Comment (gaaclarke) [inline third_party/txt/src/skia/paragraph_builder_skia.cc]: grammar

File: third_party/txt/src/skia/paragraph_builder_skia.cc [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+const TextStyle& ParagraphBuilderSkia::PeekStyle() {
+  return txt_style_stack_.empty() ? base_style_ : txt_style_stack_.top();
+}
+
+void ParagraphBuilderSkia::AddText(const std::u16string& text) {
+  builder_->addText(text);
+}
+
+void ParagraphBuilderSkia::AddPlaceholder(PlaceholderRun& span) {
+  // TODO: implement
```
Comment (gaaclarke) [inline third_party/txt/src/skia/paragraph_builder_skia.cc]: Should we have an assert(false) or log statement in the meantime?

File: third_party/txt/src/skia/paragraph_builder_skia.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+#ifndef LIB_TXT_SRC_PARAGRAPH_BUILDER_SKIA_H_
+#define LIB_TXT_SRC_PARAGRAPH_BUILDER_SKIA_H_
+
+#include "txt/paragraph_builder.h"
+
+#include "third_party/skia/modules/skparagraph/include/ParagraphBuilder.h"
+
+namespace txt {
+
+class ParagraphBuilderSkia : public ParagraphBuilder {
```
Comment (gaaclarke) [inline third_party/txt/src/skia/paragraph_builder_skia.h]: Add docstring please.

File: third_party/txt/src/skia/paragraph_skia.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+#ifndef LIB_TXT_SRC_PARAGRAPH_SKIA_H_
+#define LIB_TXT_SRC_PARAGRAPH_SKIA_H_
+
+#include "txt/paragraph.h"
+
+#include "third_party/skia/modules/skparagraph/include/Paragraph.h"
+
+namespace txt {
+
+class ParagraphSkia : public Paragraph {
```
Comment (gaaclarke) [inline third_party/txt/src/skia/paragraph_skia.h]: Add class docstring please.

File: third_party/txt/src/txt/paragraph_builder_txt.cc [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+void ParagraphBuilderTxt::PushStyle(const TextStyle& style) {
+  size_t style_index = runs_.AddStyle(style);
+  style_stack_.push_back(style_index);
+  runs_.StartRun(style_index, text_.size());
+}
+
+void ParagraphBuilderTxt::Pop() {
+  if (style_stack_.empty())
```
Comment (gaaclarke) [inline third_party/txt/src/txt/paragraph_builder_txt.cc]: Google style requires curly braces.  From the looks of our code that isn't the rules for chromium?

File: third_party/txt/src/txt/paragraph_builder_txt.h [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+#ifndef LIB_TXT_SRC_PARAGRAPH_BUILDER_TXT_H_
+#define LIB_TXT_SRC_PARAGRAPH_BUILDER_TXT_H_
+
+#include "paragraph_builder.h"
+#include "styled_runs.h"
+
+namespace txt {
+
+class ParagraphBuilderTxt : public ParagraphBuilder {
```
Comment (gaaclarke) [inline third_party/txt/src/txt/paragraph_builder_txt.h]: Add class docstring please.

File: third_party/txt/tests/txt_test_utils.cc [hunk 10]
```diff
@@ -116,4 +116,10 @@ std::shared_ptr<FontCollection> GetTestFontCollection() {
   return collection;
 }
 
+std::unique_ptr<ParagraphTxt> BuildParagraph(
+    txt::ParagraphBuilderTxt& builder) {
+  return std::unique_ptr<txt::ParagraphTxt>(
+      static_cast<txt::ParagraphTxt*>(builder.Build().release()));
```
Comment (gaaclarke) [inline third_party/txt/tests/txt_test_utils.cc]: Why do we need a static cast here?  Might be worth a comment?

File: lib/ui/text/paragraph_builder.cc (hunk 1 above)
Comment (jason-simmons) [reply lib/ui/text/paragraph_builder.cc]: done

File: third_party/txt/src/skia/paragraph_builder_skia.cc (hunk 2 above)
Comment (jason-simmons) [reply third_party/txt/src/skia/paragraph_builder_skia.cc]: Added a comment about this

File: third_party/txt/src/skia/paragraph_builder_skia.cc (hunk 3 above)
Comment (jason-simmons) [reply third_party/txt/src/skia/paragraph_builder_skia.cc]: removed this

File: third_party/txt/src/skia/paragraph_builder_skia.cc (hunk 4 above)
Comment (jason-simmons) [reply third_party/txt/src/skia/paragraph_builder_skia.cc]: Removed this (these lines originated from a work in progress branch developed by the Skia team)

File: third_party/txt/src/skia/paragraph_builder_skia.cc (hunk 5 above)
Comment (jason-simmons) [reply third_party/txt/src/skia/paragraph_builder_skia.cc]: done

File: third_party/txt/src/skia/paragraph_builder_skia.h (hunk 6 above)
Comment (jason-simmons) [reply third_party/txt/src/skia/paragraph_builder_skia.h]: done

File: third_party/txt/src/skia/paragraph_skia.h (hunk 7 above)
Comment (jason-simmons) [reply third_party/txt/src/skia/paragraph_skia.h]: done

File: third_party/txt/src/txt/paragraph_builder_txt.cc (hunk 8 above)
Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph_builder_txt.cc]: done

File: third_party/txt/src/txt/paragraph_builder_txt.h (hunk 9 above)
Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph_builder_txt.h]: done

File: third_party/txt/tests/txt_test_utils.cc (hunk 10 above)
Comment (jason-simmons) [reply third_party/txt/tests/txt_test_utils.cc]: done

**Comment by gaaclarke** [comment]
LGTM from my read through.  I'm not really familiar with the code, you might want someone else's opinion.

File: third_party/txt/src/txt/paragraph.h [hunk 11]
```diff
@@ -135,374 +97,42 @@ class Paragraph {
...
-
-  // Passes in the text and Styled Runs. text_ and runs_ will later be passed
-  // into breaker_ in InitBreaker(), which is called in Layout().
-  void SetText(std::vector<uint16_t> text, StyledRuns runs);
-
-  void SetParagraphStyle(const ParagraphStyle& style);
+  virtual ~Paragraph() = default;
 
-  void SetFontCollection(std::shared_ptr<FontCollection> font_collection);
+  virtual double GetMaxWidth() = 0;
```
Comment (GaryQian) [inline third_party/txt/src/txt/paragraph.h]: I would prefer if the documentation/comments regarding what each of these methods do exist in the interface definition. Without an implementation to directly back it up, these methods can quickly get ambiguous to people trying to learn this code.

Also, pointing out the existing implementations of this interface can be very helpful.

File: third_party/txt/src/txt/paragraph.h [hunk 12]
```diff
@@ -17,52 +17,14 @@
 #ifndef LIB_TXT_SRC_PARAGRAPH_H_
```
Comment (GaryQian) [inline third_party/txt/src/txt/paragraph.h]: Should the interface definitions still live in the src/txt directory? I can see arguments for both leaving it or moving it one layer up.

File: third_party/txt/src/txt/paragraph.h (hunk 11 above)
Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph.h]: done

File: third_party/txt/src/txt/paragraph.h (hunk 12 above)
Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph.h]: For now I'd like to keep this in `src/txt`.  Most of the contents of `src/txt` (`TextStyle`, `FontCollection`, etc.) will be used in both the Minikin and SkShaper based code paths.

If the code layout becomes too confusing, then in a later change we can move the Minikin-specific classes (`ParagraphTxt` and `ParagraphBuilderTxt`, plus a few support classes like `PaintRecord`) out of `src/txt`

**Comment by GaryQian** [comment]
cc @Rusino

**Pull request merged**